[SYCL][ESIMD] Add verifications that device has necessary aspects

vasilytric opened the pull request:
Provide verifications that device has necessary aspects to interact with `double` or `sycl::half` data types, if device doesn't has necessary aspects such data type will be skipped during runtime

vasilytric committed:
[SYCL][ESIMD] Add verifications that device has necessary aspects

yuriykoch commented on SYCL/ESIMD/api/functional/common.hpp:
```diff
@@ -55,6 +55,24 @@ using shared_allocator = sycl::usm_allocator<DataT, sycl::usm::alloc::shared>;
 using shared_vector = std::vector<DataT, shared_allocator<DataT>>;
 
+template <typename T>
+constexpr bool should_run_test_with_current_datatype(sycl::device &&device) {
```
```suggestion
constexpr bool should_run_test_with_current_datatype(const sycl::device &device) {
```
We are not modifying it or passing along anyway, so why limit usability to the r-values only? What do you think?

vasilytric replied:
this comment was applied [here](https://github.com/intel/llvm-test-suite/pull/839/commits/1c069b5f4a52017db43d26010feca6a451457ef9#diff-ef0766a5c874c909fd25e4eec71893c4614c91112485144e4efcdab4d856d0a0L59-R72)

yuriykoch commented on SYCL/ESIMD/api/functional/common.hpp:
```diff
@@ -55,6 +55,24 @@ using shared_allocator = sycl::usm_allocator<DataT, sycl::usm::alloc::shared>;
+          "Device does not support double precision floating point operations");
+      return false;
+    }
+  }
+  return true;
```
```suggestion
  } else {
    // Suppress compilation warnings
    static_cast<void>(device);
    return true;
  }
```
BTW, what do you think about making the function `inline`?

vasilytric replied:
this comment was applied [here](https://github.com/intel/llvm-test-suite/pull/839/commits/1c069b5f4a52017db43d26010feca6a451457ef9#diff-ef0766a5c874c909fd25e4eec71893c4614c91112485144e4efcdab4d856d0a0R84-R86)

yuriykoch commented on SYCL/ESIMD/api/functional/ctors/ctor_array.hpp:
```diff
@@ -96,6 +96,9 @@ template <typename DataT, typename SizeT, typename TestCaseT> class run_test {
 
 public:
   bool operator()(sycl::queue &queue, const std::string &data_type) {
+    if (!should_run_test_with_current_datatype<DataT>(queue.get_device())) {
+      return true;
+    }
```
```suggestion
    if (!should_run_test_with<DataT>(queue.get_device())) {
      return true;
    }
```
As a suggestion.  Or maybe even
```suggestion
    if (should_skip_test_with<DataT>(queue.get_device())) {
      return true;
    }
```

vasilytric replied:
the function was renamed [here](https://github.com/intel/llvm-test-suite/pull/839/commits/1c069b5f4a52017db43d26010feca6a451457ef9#diff-ef0766a5c874c909fd25e4eec71893c4614c91112485144e4efcdab4d856d0a0R71)

yuriykoch commented on SYCL/ESIMD/api/functional/common.hpp:
```diff
@@ -55,6 +55,24 @@ using shared_allocator = sycl::usm_allocator<DataT, sycl::usm::alloc::shared>;
 template <typename DataT>
 using shared_vector = std::vector<DataT, shared_allocator<DataT>>;
 
+template <typename T>
+constexpr bool should_run_test_with_current_datatype(sycl::device &&device) {
```
Would you mind to provide a comment to the function, with usage example to have API documented?
For example, for some test case depending on `SrcT` and `DstT`.

vasilytric replied:
comment was added [here](https://github.com/intel/llvm-test-suite/pull/839/commits/1c069b5f4a52017db43d26010feca6a451457ef9#diff-ef0766a5c874c909fd25e4eec71893c4614c91112485144e4efcdab4d856d0a0R58-R69)

yuriykoch commented:
Thanks, Vasily

vasilytric committed:
[SYCL][ESIMD] Structure changes

- Rename should_run_test_with_current_datatype with should_skip_test_with
- Add comment to should_skip_test_with function
- Change should_skip_test_with function signature: obtain reference to device
  instead of rvalue device

vasilytric committed:
[SYCL][ESIMD] Add should_skip_test_with to another tests

vasilytric commented:
@kbobrovs Would you mind to take a look?

vasilytric committed:
Merge branch 'intel' into verify_types

v-klochkov commented on SYCL/ESIMD/api/functional/common.hpp:
```diff
@@ -55,6 +55,40 @@ using shared_allocator = sycl::usm_allocator<DataT, sycl::usm::alloc::shared>;
+// Usage example:
+//
+// sycl::device device = queue.get_device();
+// if (should_skip_test_with(device)) {
```
Minor comment: the example usage is incomplete. The call needs the data type.
In my opinion, the lines 60-69 should be simply removed.
Having the description at the lines 58-59 is sufficient.

kbobrovs replied:
@vasilytric - please address review comments

vasilytric replied:
@kbobrovs I've applied this comment in this commit: 4977d372526f159aac3737caa810e2c99f6dc359

vasilytric committed:
[SYCL][ESIMD] Remove usage example of should_skip_test_with function

vasilytric commented:
@v-klochkov all comments has been applied, would you mind to merge this PR?
